### ISSUE TITLE
Predicted bounding boxes are all the same

### ISSUE THREAD

**Issue by RicardoQuinteiro** [facebookresearch/detr#202]
This issue is a bit odd and I am sure that I am doing something wrong. I just want some tips that might lead me to find the origin of it.

So, I have trained DETR from scratch (only with a pre-trained backbone) with num_queries=10 and 1 class. I saved several models over the epochs, however, when I am running an image through any of them I get 10 bounding boxes that are all the same and with the exact same probabilities. For models corresponding to different epochs the predicted bounding box and probability change but for every single model the 10 predicted boxes are exactly the same and the probabilities too.

Could someone give me a hint of what I could be doing wrong? I think this is really strange and it is likely a rookie mistake from my side.

**Comment by fmassa** [comment]
Hi,

Training from scratch will require a large dataset and a lot of training iterations. The issue you are seeing is because you haven't trained the model long enough. Additionally, you might not have enough data for it to generalize even if training for longer.

I would strongly recommend to first trying to fine tune the model from our pre-trained weights.

**Comment by RicardoQuinteiro** [comment]
I have done that and the results weren't great. The strategy I was trying was to train it from scratch with another dataset that has objects that have similar features to those of the object I want to detect and then use those weights to start training it with the actual dataset that I want. 

But yes, I believe I was still not using enough images to train it. How many do you reckon would be needed? And do you think it's a viable strategy?

I also have another question. Is it possible to change the neural network to receive grayscale images (with only one channel) instead? The images I have are originally in grayscale but they were converted to rgb to be used in object detection algorithms.

**Comment by fmassa** [comment]
> But yes, I believe I was still not using enough images to train it. How many do you reckon would be needed? And do you think it's a viable strategy?

We trained DETR on COCO, which has ~118k images, and we trained our models for 500 epochs (although 300 epochs already give good results).  Training on a dataset of the scale of Pascal VOC 2007+2012 can work (~15k images IIRC), although it's the lower bound of what I would expect it to work well.

> Is it possible to change the neural network to receive grayscale images (with only one channel) instead? The images I have are originally in grayscale but they were converted to rgb to be used in object detection algorithms.

I would just convert the image to be RGB, and it would make things much easier and simpler to use the pre-trained weights.

> I have done that and the results weren't great.

I would recommend having a look at https://github.com/facebookresearch/detr/issues/9 , there is a lot of useful discussion there and users have been able to train DETR on datasets with very different statistics than COCO.

I believe I have answered your questions, and as such I'm closing this issue, but let us know if you have further questions

**Comment by hellowangqian** [comment]
Hi @fmassa , would you mind sharing your experimental results on VOC dataset? For the setting with VOC 2007+2012 as the training data, do you train from scratch or from pre-trained weights on COCO?
I've done some experiments on VOC 2007 (training with trainval set ~5k images and test with test set ~ 5k images) but can only get mAP~26%. When I fine-tune the model by setting --resume option, I can get mAP~76%, both with ResNet50 as the backbone and default settings for others. I'm wondering how much difference adding VOC2012 to the training set would make. Thanks for your information.

> We trained DETR on COCO, which has ~118k images, and we trained our models for 500 epochs (although 300 epochs already give good results). Training on a dataset of the scale of Pascal VOC 2007+2012 can work (~15k images IIRC), although it's the lower bound of what I would expect it to work well.

**Comment by ducvuuit** [comment]
Hi @hellowangqian, Can you give me your email? I also train Pascal VOC on DETR but meet some problems that I don't know what.  I trained 70 epoch but mAP is always very low (~0).

**Comment by ghy0324** [comment]
Hi @fmassa, I don't think it's the reason about lack of training data.
In my experiments, the model performs reasonable on test data when use `model.train()`.
However, it just predicts same boxes and probabilities  when use `model.eval()`.

> Hi,
> 
> Training from scratch will require a large dataset and a lot of training iterations. The issue you are seeing is because you haven't trained the model long enough. Additionally, you might not have enough data for it to generalize even if training for longer.
> 
> I would strongly recommend to first trying to fine tune the model from our pre-trained weights.

**Comment by pxliang** [comment]
> Hi @fmassa, I don't think it's the reason about lack of training data. In my experiments, the model performs reasonable on test data when use `model.train()`. However, it just predicts same boxes and probabilities when use `model.eval()`.
> 
> > Hi,
> > Training from scratch will require a large dataset and a lot of training iterations. The issue you are seeing is because you haven't trained the model long enough. Additionally, you might not have enough data for it to generalize even if training for longer.
> > I would strongly recommend to first trying to fine tune the model from our pre-trained weights.

Yes, I met the same issue. Have you figured it out?

**Comment by mariiak2021** [comment]
Hi @pxliang , @ghy0324 , @fmassa can use explain about the issue with .train() and eval() further? i will appreciate any information as I have the same problem now with my network trained on a custom dataset.

**Comment by PressEtoRace** [comment]
> 

I met the same issue, is there any solutions? thanks!
